File: baseandroid/src/main/java/com/ocse/baseandroid/result/AvoidOnResult.java
package com.ocse.baseandroid.result;

import android.content.Intent;

import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import io.reactivex.Observable;

/**
 * @author hujiayi
 */
public class AvoidOnResult {
    private static final String TAG = "AvoidOnResult";
    private AvoidOnResultFragment mAvoidOnResultFragment;

    public AvoidOnResult(FragmentActivity activity) {
        mAvoidOnResultFragment = getAvoidOnResultFragment(activity);
    }


    private AvoidOnResultFragment getAvoidOnResultFragment(FragmentActivity activity) {
        AvoidOnResultFragment avoidOnResultFragment = findAvoidOnResultFragment(activity);
        if (avoidOnResultFragment == null) {
            avoidOnResultFragment = new AvoidOnResultFragment();
            FragmentManager fragmentManager = activity.getSupportFragmentManager();
            fragmentManager
                    .beginTransaction()
                    .add(avoidOnResultFragment, TAG)
                    .commitAllowingStateLoss();
            fragmentManager.executePendingTransactions();
        }
        return avoidOnResultFragment;
    }

    private AvoidOnResultFragment findAvoidOnResultFragment(FragmentActivity activity) {
        return (AvoidOnResultFragment) activity.getSupportFragmentManager().findFragmentByTag(TAG);
    }

    public Observable<ActivityResultInfo> startForResult(Intent intent) {
        return mAvoidOnResultFragment.startForResult(intent);
    }

    public Observable<ActivityResultInfo> startForResult(Class<?> clazz) {
        Intent intent = new Intent(mAvoidOnResultFragment.getActivity(), clazz);
        return startForResult(intent);
    }

    public void startForResult(Intent intent, Callback callback) {
        mAvoidOnResultFragment.startForResult(intent, callback);
    }

    public void startForResult(Class<?> clazz, Callback callback) {
        Intent intent = new Intent(mAvoidOnResultFragment.getActivity(), clazz);
        startForResult(intent, callback);
    }

    public interface Callback {
        void onActivityResult(int resultCode, Intent data);
    }
}

File: baseandroid/src/main/java/com/ocse/baseandroid/result/AvoidOnResult.kt
package com.ocse.baseandroid.result

import android.content.Intent
import androidx.fragment.app.FragmentActivity
import io.reactivex.Observable

/**
 * @author hujiayi
 */
class AvoidOnResult(activity: FragmentActivity) {
    private val mAvoidOnResultFragment: AvoidOnResultFragment
    private fun getAvoidOnResultFragment(activity: FragmentActivity): AvoidOnResultFragment {
        var avoidOnResultFragment = findAvoidOnResultFragment(activity)
        if (avoidOnResultFragment == null) {
            avoidOnResultFragment = AvoidOnResultFragment()
            val fragmentManager =
                activity.supportFragmentManager
            fragmentManager
                .beginTransaction()
                .add(avoidOnResultFragment, TAG)
                .commitAllowingStateLoss()
            fragmentManager.executePendingTransactions()
        }
        return avoidOnResultFragment
    }

    private fun findAvoidOnResultFragment(activity: FragmentActivity): AvoidOnResultFragment? {
        return activity.supportFragmentManager
            .findFragmentByTag(TAG) as AvoidOnResultFragment?
    }

    fun startForResult(intent: Intent?): Observable<ActivityResultInfo> {
        return mAvoidOnResultFragment.startForResult(intent)
    }

    fun startForResult(clazz: Class<*>?): Observable<ActivityResultInfo> {
        val intent = Intent(mAvoidOnResultFragment.activity, clazz)
        return startForResult(intent)
    }

    fun startForResult(
        intent: Intent?,
        callback: Callback?
    ) {
        mAvoidOnResultFragment.startForResult(intent, callback)
    }

    fun startForResult(
        clazz: Class<*>?,
        callback: Callback?
    ) {
        val intent = Intent(mAvoidOnResultFragment.activity, clazz)
        startForResult(intent, callback)
    }

    interface Callback {
        fun onActivityResult(resultCode: Int, data: Intent?)
    }

    companion object {
        private const val TAG = "AvoidOnResult"
    }

    init {
        mAvoidOnResultFragment = getAvoidOnResultFragment(activity)
    }
}